doc/source/scipeps/1-Optimizer.rst

```diff
@@ -0,0 +1,981 @@
+
+    class Quad(Function):
+        def __init__(self, bkg):
+            super(Quad, self).__init__(self)
+            self.bkg = bkg
+
+        def func(self, x):
+            return (x**2 + args[0])
+
+        def grad(self, x):
+            return 2*x
+
+        def hess(self, x):
+            return 2
+
+    opt = BFGS(Quad, x0).solve(maxiter=10)
+
+This is an example of machine learning. A function (``L2Loss``) is defined and
+needs to be minimized over different training examples.
+
+.. code-block:: python
+
+    from scipy.optimize import Function, Optimizer
+
+	# create your own GradientDescent optimizer, minimizing L2norm
+    class L2Loss(Function):
+        def __init__(self, A, y, *args, **kwargs):
+            self.A = A
+            self.y = y
+            super().__init__(self, *args, **kwargs)
+
+        def func(x):
+            return LA.norm(self.A@x - self.y)**2
+
+        def grad(x):
+            return 2 * self.A.T @ (self.A@x - self.y)
+
+    class GradientDescent(Optimizer):
+        def __init__(self, *args, step_size=1e-3, **kwargs):
+            self.step_size = step_size
+            super().__init__(*arg, **kwargs)
+
+        def __next__(self):
+            self.x -= self.step_size*self.grad(x)
+
+	n, d = 100, 10
+	A = np.random.randn(n, d)
+	x_star = np.random.randn(d)
+	y = np.sign(A @ x_star)
+
+	loss = L2Loss(A, y)
+	opt = GradientDescent(loss)
+
+	for k, _ in enumerate(opt):  # Optimizer.__next__ implement minimization
+		if k % 100 == 0:
+			compute_stats(opt, loss)
+
+
+Context managers offer the chance for cleanup actions, for example multiprocessing:
+
+.. code-block:: python
+   
+   with DifferentialEvolutionSolver(function, bounds, workers=2) as opt:
+        # the __entry__ and __exit__ in the solver could create and close
+        # multiprocessing pools.
+        res = opt.solve()
+
+Other possible scipy.optimize improvements
+------------------------------------------
+
+These suggestions aren't directly related to the PEP, but are worth
+mentioning to provide a picture of other possible ``optimize``
+improvements.
+
+1. Cythonise scalar minimizers, and allow ``LowLevelCallable``s as the
+   objective function. ``approx_derivative`` and ``approx_fprime`` (and
+   other support functions) would have to be ported. The real speedup
+   would not be Cythonisation per-se, but the use of `LowLevelCallable`
+   from Cython, avoiding Python call overhead. Experience gained during
+   this project might aid application in ``leastsq``, if
+   ``LowLevelCallables`` could be integrated with its Fortran nature.
+2. Implement (optional) parallelisation in calculation of finite
+   difference gradients/Hessians (e.g. ``approx_derivative``).
+3. Use parallelisation in `optimize.brute`
+4. Use parallelisation in NelderMead (10.1007/s10614-007-9094-2, http://www.econ.nyu.edu/user/wiswall/research/lee_wiswall_parallel_simplex_edit_2_8_2007.pdf).
+
+Copyright
+=========
+Andrew Nelson and Scott Sievert, Feb 2018.
```

Comment:
May I suggest that for all SciPEPs we use "This document has been placed in the public domain." (like for PEPs)?